src/node_file-inl.h

```diff
@@ -86,13 +86,13 @@ template <typename NativeT, typename V8T>
 void FillStatsArray(AliasedBufferBase<NativeT, V8T>* fields,
                     const uv_stat_t* s,
                     const size_t offset) {
-#define SET_FIELD_WITH_STAT(stat_offset, stat)                               \
-  fields->SetValue(offset + static_cast<size_t>(FsStatsOffset::stat_offset), \
+#define SET_FIELD_WITH_STAT(stat_offset, stat)                                 \
+  fields->SetValue(offset + static_cast<size_t>(FsStatsOffset::stat_offset),   \
                    static_cast<NativeT>(stat))
 
-#define SET_FIELD_WITH_TIME_STAT(stat_offset, stat)                          \
-  /* NOLINTNEXTLINE(runtime/int) */                                          \
-  SET_FIELD_WITH_STAT(stat_offset, static_cast<unsigned long>(stat))
+#define SET_FIELD_WITH_TIME_STAT(stat_offset, stat)                            \
+  /* NOLINTNEXTLINE(runtime/int) */                                            \
+  SET_FIELD_WITH_STAT(stat_offset, static_cast<double>(stat))
```

Comment:
This may bring back #12607, #13255 and/or the infamous #12115. They are why the cast to `unsigned long` was introduced, IIRC.

Reply:
After reading those fully:

- https://github.com/nodejs/node/pull/12607 mainly has discussion about `Date` rounding which is a bit hacky and relied on non-negative numbers. I adjusted that part although it can be refactored into something else.
- https://github.com/nodejs/node/issues/13255 as mentioned above, I can reproduce precision loss from this issue on v18.4.0. On top of that, I can reproduce much worse results on 32bit build. I don't see how cast to `unsigned long` helps.
- https://github.com/nodejs/node/issues/12115 is unrelated: it's about `st_ino` collisions which are not affected by this particular cast.

Overall it seems that precision loss is inevitable and accepted, and the solution is to simply use `{bigint:true}` whenever we deal with very precise and/or very large dates.

I've ran a dumb test to see precision regression between `long` and `double` and didn't see any difference (however, I'll definitely have to exclude borderline values from proper test 😅).
https://ci.nodejs.org/job/node-stress-single-test/352/ - `double`
https://ci.nodejs.org/job/node-stress-single-test/353/ - `long`

Of course if I'm missing something else, it can be reverted to `long`. The only reason why I prefer `double` is that these values will end up being interptered as 64-bit floats on JS side.

Reply:
You were right, it did indeed bring back "overflow" from https://github.com/nodejs/node/issues/13255 on win32.
Adjusted the code so it only affects other platforms now.

Any chance for this `long` to become longer? I'd expect `int64_t` or at least `time_t` but not sure how breaking it would be.
https://github.com/libuv/libuv/blob/c4d73c306b87e32acba3a77f933da72ac5604659/src/win/fs.c#L100

Reply:
Breaking, I'm afraid. `long` was the wrong choice even back in 2013<sup>1</sup>, when `uv_timespec_t` was introduced, but libuv can't change that without breaking ABI. How problematic is this for you?

<sup>1</sup> Gah, I even [pointed it out](https://github.com/joyent/libuv/pull/739#discussion_r3340950) but this particular instance slipped through.

Reply:
It leaves us with a choice on windows: we either support negative half of int32, or upper half of uint32. Y2038 is pretty close and we have an explicit test that it works:
https://github.com/nodejs/node/blob/acd2a3210de92f325e95b4b5edc304f8e70b4063/test/parallel/test-fs-utimes-y2K38.js#L45-L48
In current state of this PR, it fixes the issue for all platforms except windows, which is IMO the best solution for now.

On a larger scale, if Node.js will be ready to migrate to newer ABI, I'd like to see more appropriate types. In JS part, I think it would be reasonable to drop `Date` quirks in favor of `Temporal`: this will mean switching from ms to ns as "default" units, making bigints more necessary and `double`/`Float64` less accurate.

Reply:
I completely forgot but I tried fixing pretty much the exact same issue in pretty much the exact same way in #32408... and looks like I ran into pretty much the exact same issues as you do now. 🤦 

You can add a `Fixes: #32369` if/when this gets merged. That issue got closed by accident because I added a `Fixes:` tag to the libuv commit. Again 🤦

Reply:
Huh, I guess searching for previous issues could save quite a time here. 😅